agent/grpc-external/services/resource/watch_test.go

```diff
@@ -0,0 +1,216 @@
+	require.Contains(t, err.Error(), "resource type mesh/v1/service not registered")
+}
+
+func TestWatchList_GroupVersionMatches(t *testing.T) {
+	ctx := testContext(t)
+	registry := resource.NewRegistry()
+	registry.Register(resource.Registration{Type: typev1})
+	backend, err := inmem.NewBackend()
+	require.NoError(t, err)
+	go backend.Run(ctx)
+	server := NewServer(Config{registry: registry, backend: backend})
+	client := testClient(t, server)
+
+	// create a watch
+	stream, err := client.WatchList(ctx, &pbresource.WatchListRequest{
+		Type:       typev1,
+		Tenancy:    tenancy,
+		NamePrefix: "",
+	})
+	require.NoError(t, err)
+	rspCh := handleResourceStream(t, stream)
+
+	// insert and verify upsert event received
+	r1, err := backend.WriteCAS(ctx, resourcev1)
+	require.NoError(t, err)
+	rsp := mustGetResource(t, rspCh)
+	require.Equal(t, pbresource.WatchListResponse_OPERATION_UPSERT, rsp.Operation)
+	prototest.AssertDeepEqual(t, r1, rsp.Resource)
+
+	// update and verify upsert event received
+	r2 := clone(r1)
+	r2, err = backend.WriteCAS(ctx, r2)
+	require.NoError(t, err)
+	rsp = mustGetResource(t, rspCh)
+	require.Equal(t, pbresource.WatchListResponse_OPERATION_UPSERT, rsp.Operation)
+	prototest.AssertDeepEqual(t, r2, rsp.Resource)
+
+	// delete and verify delete event received
+	err = backend.DeleteCAS(ctx, r2.Id, r2.Version)
+	require.NoError(t, err)
+	rsp = mustGetResource(t, rspCh)
+	require.Equal(t, pbresource.WatchListResponse_OPERATION_DELETE, rsp.Operation)
+}
+
+func TestWatchList_GroupVersionMismatch(t *testing.T) {
+	// Given a watch on typev2 that only differs from typev1 by GroupVersion
+	// When a resource of typev1 is created/updated/deleted
+	// Then no watch events should be emitted
+	registry := resource.NewRegistry()
+	registry.Register(resource.Registration{Type: typev1})
+	registry.Register(resource.Registration{Type: typev2})
+	backend, err := inmem.NewBackend()
+	require.NoError(t, err)
+	ctx := testContext(t)
+	go backend.Run(ctx)
+	server := NewServer(Config{registry: registry, backend: backend})
+	client := testClient(t, server)
+
+	// create a watch for typev2
+	stream, err := client.WatchList(ctx, &pbresource.WatchListRequest{
+		Type:       typev2,
+		Tenancy:    tenancy,
+		NamePrefix: "",
+	})
+	require.NoError(t, err)
+	rspCh := handleResourceStream(t, stream)
+
+	// insert
+	r1, err := backend.WriteCAS(ctx, resourcev1)
+	require.NoError(t, err)
+
+	// update
+	r2 := clone(r1)
+	r2, err = backend.WriteCAS(ctx, r2)
+	require.NoError(t, err)
+
+	// delete
+	err = backend.DeleteCAS(ctx, r2.Id, r2.Version)
+	require.NoError(t, err)
+
+	// verify no events received
+	mustGetNoResource(t, rspCh)
+}
+
+func mustGetNoResource(t *testing.T, ch <-chan resourceOrError) {
+	t.Helper()
+
+	select {
+	case rsp := <-ch:
+		require.NoError(t, rsp.err)
+		require.Nil(t, rsp.rsp, "expected nil response with no error")
+	case <-time.After(1 * time.Second):
+		// Dan: Any way to do this without the annoying delay?
```

Comment:
Nope, sorry! 🙈

We can probably make this delay shorter (like 250ms maybe) or make it configurable using [`-short` flag](https://pkg.go.dev/testing#Short).

It might also help to mark all of these tests as [parallel](https://pkg.go.dev/testing#T.Parallel) to avoid it holding up the suite.

Reply:
👍 Fixed in https://github.com/hashicorp/consul/pull/16726/commits/ed4564807dbcb090eed4b2ddf3db9768ac698171